File: backend/src/bin/fileTree.cpp
#include <iostream>
#include <filesystem>
#include <string>
#include <nlohmann/json.hpp>
#include <stdexcept>

namespace fs = std::filesystem;
using json = nlohmann::json;

const int MAX_DEPTH = 5;

bool shouldIgnore(const std::string& name) {
    static const std::vector<std::string> ignoreList = {
        "node_modules", ".git", "dist", "build", ".next", "coverage", ".cache"
    };
    return std::find(ignoreList.begin(), ignoreList.end(), name) != ignoreList.end();
}

json buildTree(const fs::path& path, int depth = 0) {
    if (depth > MAX_DEPTH || !fs::exists(path)) return json::array();

    json children = json::array();

    try {
        for (const auto& entry : fs::directory_iterator(path)) {
            std::string name = entry.path().filename().string();
            if (shouldIgnore(name)) continue;

            json node;
            node["name"] = name;
            node["path"] = entry.path().string();
            node["depth"] = depth;

            if (entry.is_directory()) {
                node["type"] = "folder";
                node["children"] = buildTree(entry.path(), depth + 1);
            } else {
                node["type"] = "file";
            }

            children.push_back(node);
        }
    } catch (const std::exception& e) {
        std::cerr << "Error reading directory: " << e.what() << std::endl;
    }

    return children;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: ./filetree <directory>" << std::endl;
        return 1;
    }

    fs::path root(argv[1]);
    if (!fs::exists(root)) {
        std::cerr << "Directory does not exist." << std::endl;
        return 1;
    }

    json result = buildTree(root);
    std::cout << result.dump();
    return 0;
}

File: backend/src/utils/fileTree.js
import fs from 'fs';
import path from 'path';

const IGNORED_NAMES = new Set([
  '.git', '.gitignore', 'node_modules', 'dist', 'build', '.env',
  'package-lock.json', 'yarn.lock', '__tests__', '.DS_Store', 'coverage',
]);

const isIgnored = (name) => {
  return IGNORED_NAMES.has(name) || name.endsWith('.test.js') || name.endsWith('.spec.js');
};

const buildFileTree = (dir, depth = 0, maxDepth=4) => {
  const result = [];

  const items = fs.readdirSync(dir);
  for (const item of items) {
    if (isIgnored(item)) continue;

    const fullPath = path.join(dir, item);
    const stat = fs.statSync(fullPath);
    const isDir = stat.isDirectory();

    const node = {
      name: item,
      type: isDir ? 'folder' : 'file',
      path: fullPath,
      depth,
    };

    if (isDir && depth < maxDepth) {
      node.children = buildFileTree(fullPath, depth + 1, maxDepth);
    }

    result.push(node);
  }

  return result;
};

export { buildFileTree };
